### DISCUSSION TITLE
Porter Golang library Credentials

### DISCUSSION THREAD

**Discussion by martin-schoenegger** [getporter/porter#1677] [Q&A]
Hello, good evening.

I am trying to use `porter.InstallBundle` from the porter golang library ONLY with provided credentials without the need for being logged in with docker to install a bundle.
Is this currently possible? If so, is there somewhere a minimal working example available?

This is my current implementation which works perfectly fine when logged in with `docker login my-registry`

```
var p *porter.Porter

func InstallBundle(ctx context.Context, packageName string, registry string, version semver.Version) error {
    registryBundleUrl := fmt.Sprintf("%s/%s:v%s", registry, packageName, version.String())
    installOpts := porter.NewInstallOptions()
    installOpts.Name = packageName
    installOpts.Reference = registryBundleUrl
    installOpts.BundlePullOptions = porter.BundlePullOptions{Reference: registryBundleUrl, InsecureRegistry: false, Force: false}
    installOpts.AllowAccessToDockerHost = true
   // The credentialsName is added here and the debug log correctly logs them during execution
   // params: [porter-debug]
   // creds: [Username Password]
    installOpts.CredentialIdentifiers = append(installOpts.CredentialIdentifiers, credentialsName)
	
    err := installOpts.Validate([]string{packageName}, p)
    if err != nil {
        return err
    }

    err = p.InstallBundle(installOpts)
    if err != nil {
	return err
    }
    return nil

}
```
The secret is created like this:
I assume the error to be here.
```
registryCredentials := credentials.NewCredentialSet(credentialsName,
               valuesource.Strategy{
			Name: "Username",
			Source: valuesource.Source{
				Key:   "value",
				Value: "username-value",
			},
		},
		valuesource.Strategy{
			Name: "Password",
			Source: valuesource.Source{
				Key:   "value",
				Value: "password-value",
			},
		},
	)

	err := porter.Credentials.Validate(registryCredentials)
	if err != nil {
	       // handle error
        }
	porter.Credentials.Save(registryCredentials)
```
Thank you in advance for your advice and comments.

Ps: I would like to provide fully working examples for the porter golang library as I only found very few during my learning path.

Best regards, Martin

**Comment by carolynvs** [comment]
Hi Martin! Sounds like you are trying to do neat stuff with Porter. 👍 

I was wondering if you could explain a bit about what you are trying to do at a higher level? Sounds like perhaps running a docker image isn't available on the environment where you are running the bundle?

Just a quick warning that porter build, and porter install/upgrade/uninstall all require being able to run a docker container (because bundles _are_ docker images).

**Reply by martin-schoenegger** [reply]
Thank you for coming back to me @carolynvs . I am trying to build an update-service which is deployed in many different k8s clusters which are geographically separated.  Basically a similar solution as the `getporter/operator` which you are currently working on if I am not wrong. The idea is to fill the gap until this project is available. 
Ideally I would like to be able to retrieve the credentials to a corporate registry from a k8s secret and then perform `porter.InstallBundle()` (or any other porter command that requires to interact with a registry) in order to bring the cluster to a desired state.

**Reply by carolynvs** [reply]
Yup! I am knee-deep in making the necessary changes to Porter to support the operator. It may not look like things are moving forward with it right now because the changes are actually happening on the v1 branch of Porter, but it's my number one priority right now.

I think what you will want to do is use some of the same tricks that the porter operator is doing. I will follow up with a diagram of what is executed where later today, but here's the key parts:

1. the operator runs on the k8s cluster and monitors for new work.
2. Before running a porter command, the operator creates a volume that porter and the bundle invocation image (installer) will use to share data between each other.
3. the operator runs a k8s job that executes the porter command. It uses flags and environment variables to change porter's default behavior so that porter uses the kubernetes driver to execute the bundle (with the `--driver=kubernetes` flag). Here is a link to the environment variables that the operator sets https://github.com/getporter/operator/blob/56046df963fe6e1bda7787f9d9e237ccfb616600/controllers/installation_controller.go#L213.
4. the invocation image (installer) is executed in another job on the cluster. The porter job waits for it to finish, and cleans up the shared volume, before recording the result of the run.

I recommend using the [kubernetes plugin](https://porter.sh/plugins/kubernetes/) which stores porter's data on the cluster directly as k8s secrets only for light usage. Clusters do not operate well when the number of secrets gets large. For production use, select a plugin such as the azure plugin to store data remotely.

If you tried using the operator, please let me know if you had any feedback or can share why it doesn't work for you right now. That will help me improve the operator and ensure that it meets your needs when it is ready.

**Reply by martin-schoenegger** [reply]
Thanks a lot for your input. I really appreciate it and I will check out the k8s plugin.
